src/ekam/rules/intercept.c

```diff
@@ -285,6 +286,103 @@ static void canonicalizePath(char* path) {
   }
 }
 
+static bool path_has_prefix(const char* pathname, const char* prefix, size_t prefix_length) {
+  // Returns true if pathname is a child within prefix or if it matches prefix directly. Prefix must
+  // always end with a trailing '/'.
+  assert(prefix_length > 1);
+  assert(prefix_length <= PATH_MAX);
+  assert(prefix[prefix_length - 1] == '/');
+  if (strncmp(pathname, prefix, prefix_length - 1) == 0) {
+    return pathname[prefix_length - 1] == '\0' || pathname[prefix_length - 1] == '/';
+  }
+
+  return false;
+}
+
+static bool bypass_remap(const char *pathname) {
+  int debug = EKAM_DEBUG;
+
+  if (pathname[0] != '/') {
+    // Only absolute paths are allowed to bypass the remap.
+    return false;
+  }
+
+  char* bypass_dirs = getenv("EKAM_REMAP_BYPASS_DIRS");
+  if (bypass_dirs == NULL) {
+    return false;
+  }
+
+  bool found = false;
+
+  bypass_dirs = strdup(bypass_dirs);
+  // Duplicated here because I want to avoid any concurrency risk with multiple threads modifying
+  // the environment variable to do in-situ processing (& I don't want to acquire a lock).
+
+  while (bypass_dirs && *bypass_dirs) {
+    char* separator_location = strchr(bypass_dirs, ':');
+    size_t bypass_dir_length;
+    if (separator_location) {
+      *separator_location = '\0';
+      bypass_dir_length = separator_location - bypass_dirs;
+    } else {
+      bypass_dir_length = strlen(bypass_dirs);
+    }
+    assert(bypass_dirs[bypass_dir_length] == '\0');
+
+    if (bypass_dirs[0] != '/') {
+      fprintf(stderr, "Bypass directory '%s' isn't an absolute path. Skipping.\n", bypass_dirs);
+    } else if (bypass_dirs[bypass_dir_length - 1] != '/') {
+      fprintf(stderr, "Bypass directory '%s' doesn't end in trailing '/'. Skipping.\n", bypass_dirs);
+    } else if (path_has_prefix(pathname, bypass_dirs, bypass_dir_length)) {
+      if (debug) {
+        fprintf(stderr, "Bypass found for %s\n", pathname);
+      }
+      found = true;
+      break;
+    }
+
+    if (separator_location) {
+      bypass_dirs = separator_location + 1;
+    } else {
+      bypass_dirs = NULL;
+    }
+  }
+
+  free(bypass_dirs);
```

Comment:
The code above modifies `bypass_dirs` so at best this is `free(NULL)` and you're leaking memory, while at worst you're passing a pointer to `free()` that didn't come from `malloc()` which would be UB.

Reply:
Good catch. I had run Ekam itself under ASAN but forgot that intercept.so needs some special handling. I'll make sure to retest with building intercept.so under asan/tsan to avoid surprises.